File: ios/Reguerta/Reguerta/Data/ShiftCoverage/LocalCoverageRehearsalAccess.swift
#if DEBUG
import Foundation

/// Memory-only emulator credentials. Login never touches the application's Firebase Auth instance.
@MainActor
final class LocalCoverageRehearsalAccess: CoverageRehearsalAccess {
    let repository: any ShiftCoverageRepository
    private let localRepository: LocalShiftCoverageRepository
    private let credentials: CoverageRehearsalCredentials
    private let loader: any HTTPDataLoading
    private let endpoint: URL

    init(port: Int = 8799, loader: any HTTPDataLoading = CoverageLoopbackDataLoader()) throws {
        let credentials = CoverageRehearsalCredentials()
        self.credentials = credentials
        self.loader = loader
        guard let endpoint = URL(string:
            "http://127.0.0.1:9098/identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=fake-api-key"
        ) else { throw ShiftCoverageFailure.localOnly }
        self.endpoint = endpoint
        localRepository = try LocalShiftCoverageRepository(
            port: port,
            currentSession: { credentials.session },
            tokenProvider: {
                guard let token = credentials.token else { throw ShiftCoverageFailure.sessionChanged }
                return token
            },
            loader: loader
        )
        repository = localRepository
    }

    func signOut() {
        credentials.revision &+= 1
        credentials.token = nil
        credentials.session = nil
    }

    func signIn(email: String, password: String) async throws -> ShiftCoverageSession {
        signOut()
        let owner = credentials.revision
        do {
            var request = URLRequest(url: endpoint, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 15)
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONEncoder().encode(Login(email: email, password: password))
            let (data, response) = try await loader.data(for: request)
            try Task.checkCancellation()
            guard owner == credentials.revision else { throw ShiftCoverageFailure.sessionChanged }
            guard let response = response as? HTTPURLResponse, response.url == endpoint, response.statusCode == 200,
                  let login = try? JSONDecoder().decode(LoginResult.self, from: data) else {
                throw ShiftCoverageFailure.rejected(status: 401, code: "coverage_sign_in_failed")
            }
            let provisional = ShiftCoverageSession(
                uid: login.localId,
                memberId: "resolving",
                authorizationRevision: owner
            )
            credentials.token = login.idToken
            credentials.session = provisional
            let snapshot = try await localRepository.resolveMember(session: provisional)
            guard owner == credentials.revision else { throw ShiftCoverageFailure.sessionChanged }
            let session = ShiftCoverageSession(
                uid: login.localId,
                memberId: snapshot.memberId,
                authorizationRevision: owner
            )
            credentials.session = session
            return session
        } catch {
            if owner == credentials.revision {
                signOut()
            }
            throw error
        }
    }

    private struct Login: Encodable {
        let email: String
        let password: String
        let returnSecureToken = true
    }

    private struct LoginResult: Decodable {
        let localId: String
        let idToken: String
    }
}

@MainActor
private final class CoverageRehearsalCredentials {
    var revision: UInt64 = 0
    var token: String?
    var session: ShiftCoverageSession?
}
#endif

File: android/Reguerta/app/src/debug/java/com/reguerta/user/data/shiftcoverage/LocalCoverageRehearsalAccess.kt
package com.reguerta.user.data.shiftcoverage

import androidx.annotation.MainThread
import com.reguerta.user.domain.shiftcoverage.CoverageRehearsalAccess
import com.reguerta.user.domain.shiftcoverage.ShiftCoverageFailure
import com.reguerta.user.domain.shiftcoverage.ShiftCoverageSession
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/** Memory-only Auth emulator session, separate from the application's Firebase Auth instance. */
@MainThread
internal class LocalCoverageRehearsalAccess(
    port: Int = 8799,
    emulatorHost: Boolean = true,
    private val transport: CoverageHttpTransport = LocalCoverageHttpTransport(),
) : CoverageRehearsalAccess {
    private var token: String? = null
    private var session: ShiftCoverageSession? = null
    private var revision = 0L
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private val host = if (emulatorHost) "10.0.2.2" else "127.0.0.1"
    override val repository = LocalShiftCoverageRepository(
        port = port,
        emulatorHost = emulatorHost,
        currentSession = { session },
        tokenProvider = { token ?: throw ShiftCoverageFailure.SessionChanged },
        transport = transport,
    )

    override fun signOut() {
        revision++
        token = null
        session = null
    }

    override suspend fun signIn(email: String, password: String): ShiftCoverageSession {
        signOut()
        val owner = revision
        try {
            val response = transport.post(
                url = "http://$host:9098/identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=fake-api-key",
                token = "",
                body = json.encodeToString(Login(email, password)),
            )
            currentCoroutineContext().ensureActive()
            if (owner != revision) throw ShiftCoverageFailure.SessionChanged
            val login = runCatching { json.decodeFromString<LoginResult>(response.body) }.getOrNull()
            if (response.status != 200 || login == null) {
                throw ShiftCoverageFailure.Rejected(401, "coverage_sign_in_failed")
            }
            val provisional = ShiftCoverageSession(login.localId, "resolving", owner)
            token = login.idToken
            session = provisional
            val snapshot = repository.resolveMember(session = provisional)
            if (owner != revision) throw ShiftCoverageFailure.SessionChanged
            return provisional.copy(memberId = snapshot.memberId).also { session = it }
        } catch (error: Exception) {
            if (owner == revision) signOut()
            throw error
        }
    }

    @Serializable
    private data class Login(val email: String, val password: String) {
        val returnSecureToken = true
    }
    @Serializable
    private data class LoginResult(val localId: String, val idToken: String)
}
